### PR TITLE
Make the docs clearer for new contributors

### PR DESCRIPTION
**Pull request by camelid**
It confused me before, so I made it extra obvious that you need to run
a script to set up your toolchain before you can build Clippy.

I also added a note so that new contributors aren't confused when
Clippy doesn't build as a result of a change in rustc's internals.

changelog: make `CONTRIBUTING.md` clearer for new contributors

### CODE REVIEW

**Comment by flip1995** [comment]
This is already described in `basics.md`. Maybe we should make it more obvious, that people that are new to Clippy should definitely take a look at that document.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -28,11 +29,11 @@ All contributors are expected to follow the [Rust Code of Conduct].
 
 ## Getting started
```
Comment (flip1995) [inline CONTRIBUTING.md:31]: Maybe something like:

```suggestion

**Note:** If this is the first time, you want to contribute to or build Clippy, you should first read the [`Basics docs`](doc/basics.md).

```

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -79,6 +85,23 @@ an AST expression). `match_def_path()` in Clippy's `utils` module can also be us
...
+`cargo test` like any other Rust project!
+
+**Note:** You may get compiler errors when building Clippy, even if you
+didn't change anything. That's because, as mentioned earlier, Clippy
+depends on rustc internals. There is [discussion][nightly-build-issue]
+on ways to make sure that rustc does not break Clippy in the future
+so that Clippy always builds.
+
+[nightly-build-issue]: https://github.com/rust-lang/rust-clippy/issues/5561
+
```
Comment (flip1995) [inline CONTRIBUTING.md:104]: I **really** don't want this here. Keeping the docs up to date is already hard enough as it is. -> DRY principle.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -28,11 +29,16 @@ All contributors are expected to follow the [Rust Code of Conduct].
...
+first read the [Basics docs](doc/basics.md). It will help you get set up
+for working on Clippy. There's also more information about building Clippy
+[below](#how-to-build-clippy).
+
+### High level approach
 
 1. Find something to fix/improve
 2. Change code (likely some file in `clippy_lints/src/`)
-3. Follow the instructions in the [Basics docs](doc/basics.md) such as running the `setup-toolchain.sh` script
+3. [Build Clippy](#how-to-build-clippy)
```
Comment (flip1995) [inline CONTRIBUTING.md:41]: This should stay the same, IMO.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -28,11 +29,16 @@ All contributors are expected to follow the [Rust Code of Conduct].
 
 ## Getting started
 
-High level approach:
+**Note:** If this is your first time contributing to Clippy, you should
+first read the [Basics docs](doc/basics.md). It will help you get set up
+for working on Clippy. There's also more information about building Clippy
+[below](#how-to-build-clippy).
```
Comment (flip1995) [inline CONTRIBUTING.md:35]:
```suggestion
**Note:** If this is your first time contributing to Clippy, you should
first read the [Basics docs](doc/basics.md). It will help you get set up
for working on Clippy. 
```

Keep important notes as short as possible. Maybe even just

```suggestion
**Note:** If this is your first time contributing to Clippy, you should
first read the [Basics docs](doc/basics.md).
```

File: CONTRIBUTING.md (hunk 2 above)
Comment (camelid) [reply CONTRIBUTING.md:104]: I think it might be helpful to move the note here to the basics docs, then, so that new contributors aren't confused by the compiler errors. Does that make sense to you?

File: CONTRIBUTING.md (hunk 4 above)
Comment (camelid) [reply CONTRIBUTING.md:35]: I'm thinking of somewhere in between your two versions:

```suggestion
**Note:** If this is your first time contributing to Clippy, you should
first read the [Basics docs](doc/basics.md). It will help you get your
environment set up.
```

File: CONTRIBUTING.md (hunk 3 above)
Comment (camelid) [reply CONTRIBUTING.md:41]: How about:

```suggestion
3. Follow the instructions in the [Basics docs](doc/basics.md) to get set up
```

File: CONTRIBUTING.md (hunk 2 above)
Comment (flip1995) [reply CONTRIBUTING.md:104]: Not really. Normally these errors are fixed within a few hours and most of the contributors don't see them. This is just unlucky timing, since we're waiting for other Clippy changes to get merged in rustc, before fixing those. 

In addition to that, the basics docs should be self contained and contain all the information needed to build Clippy. The split of these information between multiple documents was the reason the basics docs exist in the first place.

Comment (camelid) [reply CONTRIBUTING.md:104]: I just think it might be a bit confusing for people who are new clippy and don't understand tool breakages. Is there somewhere that you would feel comfortable putting it?

Comment (camelid) [reply CONTRIBUTING.md:104]: If not, I can remove it, I just thought it might be helpful

Comment (flip1995) [reply CONTRIBUTING.md]: Oh I thought this was already mentioned in `basics.md`. You can put it here: https://github.com/rust-lang/rust-clippy/blob/master/doc/basics.md#setup

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -28,11 +29,14 @@ All contributors are expected to follow the [Rust Code of Conduct].
 
 ## Getting started
 
-High level approach:
+**Note: If this is your first time contributing to Clippy, you should
+first read the [Basics docs](doc/basics.md).**
```
Comment (camelid) [inline CONTRIBUTING.md:33]: I bolded the whole thing to catch people's eyes (cc the rustc-dev-guide's getting started section).

File: CONTRIBUTING.md (hunk 2 above)
Comment (camelid) [reply CONTRIBUTING.md]: Hmm, not under https://github.com/rust-lang/rust-clippy/blob/master/doc/basics.md#building-and-testing?

Comment (camelid) [reply CONTRIBUTING.md]: Ah, did you suggest that since that part mentions how Clippy depends on internals?

Comment (flip1995) [reply CONTRIBUTING.md]: > Ah, did you suggest that since that part mentions how Clippy depends on internals?

Yes exactly. That way we can keep this note a little shorter, and it is higher up in the document ;)

File: doc/basics.md [hunk 6]
```diff
@@ -53,6 +53,14 @@ rustup-toolchain-install-master -f -n master -c rustc-dev -c llvm-tools
 rustup override set master
 ```
 
+**Note:** You may get compiler errors when building Clippy, even if you
+didn't change anything. That's because, as mentioned earlier, Clippy
+depends on rustc internals. There is [discussion][nightly-build-issue]
+on ways to make sure that rustc does not break Clippy in the future
+so that Clippy always builds.
```
Comment (camelid) [inline doc/basics.md:60]: Should we mention that this is called a tool breakage and link to forge docs?

File: doc/basics.md [hunk 7]
```diff
@@ -53,6 +53,14 @@ rustup-toolchain-install-master -f -n master -c rustc-dev -c llvm-tools
...
+
+[nightly-build-issue]: https://github.com/rust-lang/rust-clippy/issues/5561
+
```
Comment (flip1995) [inline doc/basics.md:63]:
```suggestion
_Note:_ Sometimes you may get compiler errors when building Clippy, even if you
didn't change anything. Normally those will be fixed by a maintainer in a few hours. 

```

and put it in front of the code block, so you don't need to mention the internals again.

Also the issue is not relevant for beginners.

File: doc/basics.md (hunk 6 above)
Comment (flip1995) [reply doc/basics.md:60]: No, this is different from classic tool breakages.

Comment (camelid) [reply doc/basics.md]: Hmm, I guess I didn't realize that. Are there docs that talk about the differences between classic tool breakages and clippy tool breakages?

File: doc/basics.md [hunk 8]
```diff
@@ -53,6 +53,9 @@ rustup-toolchain-install-master -f -n master -c rustc-dev -c llvm-tools
 rustup override set master
 ```
 
+_Note:_ Sometimes you may get compiler errors when building Clippy, even if you
```
Comment (camelid) [inline doc/basics.md:56]: This is a nit, but should I bold this and the other note lower down (so **Note:** instead of _Note:_) to match `CONTRIBUTING.md`?

Comment (flip1995) [reply doc/basics.md:56]: Normally we have _Note:_ in our docs. the note in the CONTRIBUTING.md file is an exception. So this is fine.,

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -15,6 +15,7 @@ All contributors are expected to follow the [Rust Code of Conduct].
 - [Contributing to Clippy](#contributing-to-clippy)
   - [Getting started](#getting-started)
     - [Finding something to fix/improve](#finding-something-to-fiximprove)
+    - [How to build Clippy](#how-to-build-clippy)
```
Comment (flip1995) [inline CONTRIBUTING.md:18]: And finally: Please remove this line and squash your commits. After that this is good to go 👍

Comment (camelid) [reply CONTRIBUTING.md]: Ah, I'm used to rust-lang/rust where bors auto-squashes :)

Comment (camelid) [reply CONTRIBUTING.md]: Squashed!

File: doc/basics.md (hunk 6 above)
Comment (flip1995) [reply doc/basics.md]: Yeah, see https://rustc-dev-guide.rust-lang.org/contributing.html#external-dependencies-subtree and the following section.

TL;DR: Clippy breakages will never appear in nightlies delivered with rustups, so they aren't user facing.

Comment (camelid) [reply doc/basics.md]: So when there's an internal breaking change, clippy is updated in rust-lang/rust, but then has to be brought over to this repository (rust-lang/rust-clippy)?

Comment (flip1995) [reply doc/basics.md]: Yes, exactly.
